### ISSUE TITLE
Strange behavior in power coefficient convergence plot

### ISSUE THREAD

**Issue by hcOnel** [turbinesFoam/turbinesFoam#274]
Hello, in the NREL5MW case, I get the following Cp convergence plot.
![convergence](https://user-images.githubusercontent.com/5435247/47887039-651cc600-de4e-11e8-8240-b01a792b48ac.png)
Just as I expect Cp to drop a little bit further around 30s mark, it suddenly jumps and converges at a higher value (unlike it is used to be). The rotor is isolated, no hub and tower introduced, dynamic stall is off. What might be causing this? (happens in both multi and single Reynolds cases)

**Comment by petebachant** [comment]
Do you have those particular case files up in a repo? I'd be curious to see what the flow field looks like when that happens.

**Comment by hcOnel** [comment]
Sure. I have uploaded the case as it is to my repo [here](https://github.com/hcOnel/NREL5MW-turbinesFoam/tree/master/nrel5mw_clean).

**Comment by petebachant** [comment]
Sure, I put the sketch and some description on a wiki page: https://github.com/turbinesFoam/turbinesFoam/wiki/Actuator-line-geometry

**Comment by hcOnel** [comment]
Thanks for sharing! By the way, did you have the chance to check the case? I suspected the possibility of outflow boundary being too close to the turbine and developing wake region reaching the outflow boundary and distorting the rotor plane due to its effect reaching the rotor plane after some time (which happens to be around 30s mark). But since the outflow boundary condition is inflowOutflow, it does not sound likely so.

**Comment by petebachant** [comment]
You have 10D downstream, which doesn't sound like it's too low, but you never know. I'd take a look at the pressure field evolution in ParaView and see what happens when the wake hits the outlet.

**Comment by hcOnel** [comment]
I have a new kind of oscillating behavior in Cp and Ct plots when I halved the mesh size in rotor area.
Case details: NREL5MW, Uinf=11.4m/s, TSR=8, cell size at the rotor area: 0.5625m (112 cells per rotor radius), dt~4.9E-03s, rotation is 0.406 degrees/time step, Courant number never exceeds 0.12 throughout the simulation, 34 elements defined in fvOptions, max element length is 0.6m.
![image](https://user-images.githubusercontent.com/5435247/48284448-8003c180-e470-11e8-98f9-75cdb8278b2e.png)
This is the first time I'm getting this kind of behavior. Before this case, every setting above kept -except the cell size at rotor area and time step doubled (coarser); I get at least large-oscillation free results (like the one I posted before). What might be causing this? An element size becoming very close to cell size? Gaussian smearing function has problems?

**Comment by hcOnel** [comment]
I've animated the rotation for approximately three periods of this jump:
https://media.giphy.com/media/8Bl3EHqO8czuH621Fu/giphy.gif
It looks like velocity and pressure values jump on certain angular positions. I have nothing other than the blades. I am really bugged...

**Comment by hcOnel** [comment]
By the way, I have run this simulation with same settings but changing the number of elements defined in the fvOptions file: 34 and 19. I get different magnitudes and shapes of this oscillations, however the period and positions remain the same:
nElements = 19
![nelements19](https://user-images.githubusercontent.com/5435247/48312225-8acb6d00-e5bc-11e8-9692-11c7e1c5bdc9.png)

nElements = 34
![nelements34](https://user-images.githubusercontent.com/5435247/48312227-8f902100-e5bc-11e8-8202-ac46d90f3490.png)

So it looks like the number of elements also have some impact on this.

**Comment by petebachant** [comment]
From the flow field GIF it looks like the blades all of a sudden perform better when one passes where the tower would be. It would be helpful to also see slices in the x--y and x--z planes. Also, does this happen with end effects turned off?

**Comment by hcOnel** [comment]
It is true for when a blade passes 6 o'clock position, but the same happens for 2 and 10 o'clock positions so I think that existence of a tower does not explain this, besides, there are no tower defined in fvOptions or topoSetDict or anywhere. I have checked them several times now.
GIF animations include in-plane (or tangential, Uy+Uz) and plane-normal (or axial, Ux) velocity distributions, but surely I will post other slices.
After you mentioned, I have solved the same case with no tip correction, and same thing happens. Below is a convergence plot with no tip correction:
![case](https://user-images.githubusercontent.com/5435247/48545669-80b8af80-e8d7-11e8-877f-fb7e485f2322.png)
It looks more like something "accumulates" over time and goes under a steep change at some point, and this has a (close to) 120 degrees period by chance. Or, since this is happening since I have refined the mesh, it might has something to do with cell size/cell alignment?
For the note, I have 36m farfield cell size and 6 levels of refinement at rotor plane (approx. 0.5625m cell size)

**Comment by hcOnel** [comment]
Hello, did anyone manage to reproduce the same result?
Since this is happening upon changing the number of cells (and cell size) on the rotor plane and element length has some effect on it, which part do you suggest checking in the code? I suspect the part which adds the force on cells, specifically the Gaussian distribution of singular forces on close cells. Which parts of which source files should I check for this?

**Comment by petebachant** [comment]
I have not had a chance to attempt to reproduce this result. From the flow field image, I would not expect the size of the Gaussian to be changing, but you can check the code for that here: https://github.com/turbinesFoam/turbinesFoam/blob/master/src/fvOptions/actuatorLineSource/actuatorLineElement/actuatorLineElement.C#L199

**Comment by daa-mec** [comment]
Hello, I have not enough time to test your case specifically right now (maybe I will try later), however I got a similar behaviour when I used a coarse mesh and no end effects (test case NREL phase VI). 

Regarding the time step, in your case the tip of the blade travels a distance of 0.44m with TSR=8 (or 0.39m  with TSR=7), assuming you have a uniform cartesian grid across the rotor seems right to me. From what I understood in the code, the epsilon projection is taken as the maximum of three values, epsilonMesh, epsilonDrag, and epsilonLift.... which sets a limit to have an epsilon at least equal to 4*DX (the epsilonMesh), if none of the other is larger. 

I will keep an eye on this.

**Comment by hcOnel** [comment]
Hi daa-mec, thanks for the feedback. To clarify, are you getting the sudden jump (my first post) or oscillatory behavior? It is true that I also get the sudden jump in coarse meshes, but the oscillations are strictly specific to finer meshes.

**Comment by daa-mec** [comment]
I did get oscillatory behaviour, unfortunately I did not keep the output files from the simulation. Perhaps I will re run the case to explore it in more details. Until now I have not get any sudden jump.

**Comment by petebachant** [comment]
Ah, yes that is a known issue (#29). Any post-processing scripts should take the last value for each time since I couldn't figure out a way to detect the final inner iteration.

**Comment by petebachant** [comment]
The PIMPLE algorithm does `nOuterCorrectors` additional iterations per time step. For all my cases, I set this to 1, which is identical to the PISO algorithm, and avoids the problem of multiple performance values written per time step.

**Comment by petebachant** [comment]
The Gaussian coeffs need to go in the individual profile data subdicts, i.e., they could be different for each foil type.

**Comment by hcOnel** [comment]
I've tried to observe the effect of Gaussian coeffs:
![untitled](https://user-images.githubusercontent.com/5435247/50210106-5e0e3080-0386-11e9-820d-b2880a19f841.png)
For x1, default values are used. I've multiplied all of them by 2 and 4 for other cases.
Oscillations disappear as factors are increased, but Cp value also increases.

**Comment by hcOnel** [comment]
[Here](https://github.com/turbinesFoam/turbinesFoam/blob/385859c3efbdc51ab6a7b2e49d003b4ad9d51da3/src/fvOptions/actuatorLineSource/actuatorLineElement/actuatorLineElement.C#L226), it looks like delta_r (local cell size) is multiplied by 2 twice:
```
        // Projection width based on local cell size (from Troldborg (2008))
        epsilonMesh = 2.0*Foam::cbrt(V[posCellI]);
        epsilonMesh *= meshFactor; // Cell could have non-unity aspect ratio
```
Troldborg suggests `epsilon = 2 * delta_r`, but isn't that 2 already the `meshFactor` variable? `cbrt` function returns cell edge size directly afaik, [here](https://www.openfoam.com/documentation/cpp-guide/html/guide-turbulence-les-delta-cuberootvol.html).

**Comment by petebachant** [comment]
That is true. In that other issue I went through the process where I determined that a `meshFactor` of 2 was best for the cases I was working with, but I fully admit that this was very ad hoc and worth exploring further.

**Comment by hcOnel** [comment]
Hello again,
I have been working for the past week on Gaussian smearing of the calculated singular forces to the cells in the mesh, whether the calculated forces are passed into the computational grid correctly or not and Gaussian epsilon's effect on the accuracy of relative velocity and angle of attack.
Currently I am examining the total force exerted on the turbine and cross-checking whether the components match or not between turbinesFoam's postProcessing output and actual force.turbine field variable in the domain.
To perform this check;
* I use a script which sums up the total force component by component for all blade elements for all blades, for each time step in turbinesFoam's postProcessing output.
* Then use ParaView's IntegrateVariables filter to sum up force.turbine field variable for all cells within the domain, component by component, similarly.
My expectation: they should match for the same time step. So far, I have done this for a single time step and the good news is, they are in a very close agreement:

![image](https://user-images.githubusercontent.com/5435247/51918189-23220280-23f2-11e9-8701-0ffc4117fea1.png)

Note that the sign difference is due to coordinate system convention and can be ignored, and also I do not use force per length values but total calculated force values at element centers.
But then I noticed that the force.turbine field variable should be in the units of ( force / cell volume ), since only then the momentum equation units are satisfied. There should be a ratio as much as the cell volume (considering the cell volume is homogeneous for all the cells where the force is smeared -I checked this-, it should be constant for all force components). Would you agree? Can there be an error there?
Hope my long post is clear. Best regards,
Huseyin

**Comment by petebachant** [comment]
You are correct that there is no division by cell volume--it is a force per unit density. We should double check the units. I believe the `actuationDiskSource` in OpenFOAM computes the overall force for all cells, then divides by the total volume of the disk for the force in each cell. But by analogy this would mean that we need to integrate the force for a given element over all the cells within its radius, which was done in one of the SOWFA development papers, and I followed their example.

**Comment by hcOnel** [comment]
Incompressible N-S equations, divided by constant density, has the units of acceleration (m/s^2).
![image](https://user-images.githubusercontent.com/5435247/52164626-52dc4d80-2705-11e9-977e-54ecac9cbbc9.png)
Hence, body force term must be per density per cell volume, only then the momentum equation is satisfied when integrated over the cells.
force / density / volume = N / (kg / m^3) / m^3 = (kg m / s^2) / (kg / m^3) / m^3 = m / s^2
I am not familiar with how body forces are passed to and treated by OpenFOAM's pimpleFoam solver, maybe force in Newtons is divided by cell volume in the solver?
Correct me if I am wrong.

**Comment by petebachant** [comment]
The momentum equation in `pimpleFoam` is 

```c++
// Solve the Momentum equation

MRF.correctBoundaryVelocity(U);

tmp<fvVectorMatrix> tUEqn
(
    fvm::ddt(U) + fvm::div(phi, U)
  + MRF.DDt(U)
  + turbulence->divDevReff(U)
 ==
    fvOptions(U)
);
fvVectorMatrix& UEqn = tUEqn.ref();

UEqn.relax();

fvOptions.constrain(UEqn);

if (pimple.momentumPredictor())
{
    solve(UEqn == -fvc::grad(p));

    fvOptions.correct(U);
}
```

Looking in the [SOWFA source code](https://github.com/NREL/SOWFA/blob/master/src/turbineModels/turbineModelsStandard/horizontalAxisWindTurbinesALM/horizontalAxisWindTurbinesALM.C#L1290) the `bodyForce` is not dependent on cell volume, though the force sums are.

Their solver's U equation is

```c++
            fvVectorMatrix UEqn
            (
                fvm::ddt(U)
              + fvm::div(phi, U)
              + turbulence->divDevReff(U)
              - turbines.force()
            );
```

Where `turbines.force` simply returns the `bodyForce`. 

It's worth investigating this further, but it will take me some time. Any help you could provide checking the consistency of both would be much appreciated!
